Refactor fabric data in database

dnltz opened the pull request:
This PR separates the fabric data from the part data by introducing a new mapping YAML file which contains all information about the parts and devices. The new mapping will can also allow to easily generate a complete database some day by iterating over each entry.
Some scripts and makefiles got updated to handle the new location of the fabric.
Additionally, fixed two minor issue in Python.

Two patches in symbiflow-arch-defs are required: https://github.com/dnltz/symbiflow-arch-defs/commits/WIP/dnltz/refactory_database

Needs to be rebased after #1534 was merged.
Closes #1475

litghost commented on prjxray/mapping.yaml:
```diff
@@ -0,0 +1,162 @@
+# part number to device, package and speed grade mapping
```
The mapping file should be part of the database, as after this is merged,  as #1476 will auto-generate this file.

More detailed thoughts:
 - We should have a mapping file per family
 - The "pins" map probably belongs somewhere else.  The pins stuff is more likely to be hand-crafted, where as the other parts of the mapping can and eventually should be auto generated.

dnltz replied:
Ah yes, moving the mapping file into the family directories will solve some issues I had.
I will undo the pins. My idea was to store these infos in same machine-readable files to automate the Makefiles a little bit. But since there are way more informations in these Makefiles, this might be done later or not.

litghost replied:
> Ah yes, moving the mapping file into the family directories will solve some issues I had.
> I will undo the pins. My idea was to store these infos in same machine-readable files to automate the Makefiles a little bit. But since there are way more informations in these Makefiles, this might be done later or not.

I'm totally fine with the pins change, but it should be seperate from the fabric stuff.

dnltz replied:
Not sure if "resources.yaml" are the correct name for these information?!

litghost commented on prjxray/mapping.yaml:
```diff
@@ -0,0 +1,162 @@
+
+---
+# device to fabric mapping
+"xc7a200t":
```
The fabric <-> device relationship should probably be in a separate file.

litghost commented:
Overall this looks like a good first iteration.  Couple changes that I feel are needed going forward. I'll also kick off the full CI on this version to see if anything breaks.

litghost commented on fuzzers/run_fuzzer.py:
```diff
@@ -313,7 +314,7 @@ def mem_convert(s):
             break
 
-    v = float(s[:-len(u)])
+    v = float(s[:-len(u)].replace(',', '.'))
```
Can you put a small comment what is going on here?

dnltz replied:
That's the traceback I get:
```
2021-01-11T18:55:34 - xc7a200tffg1156-1/001-part-yaml             -   10s: Traceback (most recent call last):
2021-01-11T18:55:34 - xc7a200tffg1156-1/001-part-yaml             -   10s:   File ".//run_fuzzer.py", line 564, in run_fuzzer
2021-01-11T18:55:34 - xc7a200tffg1156-1/001-part-yaml             -   10s:     mem = get_memory()['mem']
2021-01-11T18:55:34 - xc7a200tffg1156-1/001-part-yaml             -   10s:   File ".//run_fuzzer.py", line 356, in get_memory
2021-01-11T18:55:34 - xc7a200tffg1156-1/001-part-yaml             -   10s:     d[k] = mem_convert(v)
2021-01-11T18:55:34 - xc7a200tffg1156-1/001-part-yaml             -   10s:   File ".//run_fuzzer.py", line 316, in mem_convert
2021-01-11T18:55:34 - xc7a200tffg1156-1/001-part-yaml             -   10s:     v = float(s[:-len(u)])
2021-01-11T18:55:34 - xc7a200tffg1156-1/001-part-yaml             -   10s: ValueError: could not convert string to float: '5,4'
2021-01-11T18:55:35 - xc7a200tffg1156-1/001-part-yaml             -   11s: Warning: Killed program which should have been dead!
2021-01-11T18:55:35 - xc7a200tffg1156-1/001-part-yaml             -   11s: Finishing (User:0s System:0s).

```
and my free results:

```
daniel@desktop-daniel:~/work/symbiflow/prjxray$ free -mh
              total        used        free      shared  buff/cache   available
Mem:           31Gi       5,0Gi        20Gi       222Mi       6,0Gi        25Gi
Swap:         979Mi          0B       979Mi

```
Seems like my machine is still partly configured with German output, which outputs commas instead of dots.

litghost replied:
Sure, please add a comment.  The other option here is to override the LOCALE to avoid this.

mithro replied:
I highly recommend we explicitly set the `LOCALE` value, maybe somewhere like in the `Makefile`?

litghost commented on fuzzers/run_fuzzer.py:
```diff
@@ -447,6 +448,17 @@ def main(argv):
     assert os.path.exists(fuzzer_logdir)
 
+    # Set part information from the mapping file and set into the env
+    mapping_file = "../prjxray/mapping.yaml"
```
Given that 'XRAY_PART' is defined, you can use 'XRAY_DATABASE'/'XRAY_DATABASE_DIR'/'XRAY_DIR'.

Also I believe this file should be moved to `os.path.join(os.environ['XRAY_DATABASE_DIR'], os.environ['XRAY_DATABASE'])`, so there is that.

litghost commented on prjxray/db.py:
```diff
@@ -45,6 +47,10 @@ def __init__(self, db_root, part):
     """
         self.db_root = db_root
         self.part = part
+        mapping_file = os.path.join(
```
Rather than using `__file__`, this should be based on a path from `db_root`

litghost commented on settings/artix7_100t.sh:
```diff
@@ -31,15 +31,4 @@ export XRAY_ROI_GRID_X2="77"
 export XRAY_ROI_GRID_Y1="0"
```
Don't rename this file, or if you do, add another file like was done here: https://github.com/SymbiFlow/prjxray/blob/ba0ed9dba2f960b68e4d86e6963bbaf0c088466a/settings/spartan7.sh

dnltz replied:
That's also the problem why the CI build failed. I thought it might be more intuitive that the 100t is not hidden in the artix7.sh. Will undo this, too.

dnltz commented:
@litghost - Pushed a new, rebased version.

- Moved the mapping files into the database (https://github.com/SymbiFlow/prjxray-db/pull/15) and changed the pathes to the DB variables.
- Added an "export LC_ALL=C" to the utils/environment instead for replacing character which sets the localization settings to default.
- Undid the settings renaming.

mithro commented:
Looks like this is failing due to the `Cleaning out current database`?
```
========================================
Cleaning out current database
----------------------------------------
rm -f zynq7/*.csv zynq7/*.db zynq7/*.json zynq7/*.yaml zynq7/*.fasm
rm -f zynq7/timings/*.sdf
rm -rf zynq7/xc7*/
----------------------------------------

========================================
Running Database build
----------------------------------------
make --dry-run
source settings/zynq7.sh && make -C fuzzers
make[1]: Entering directory '/tmpfs/src/github/symbiflow-prjxray-presubmit-db-zynq7/fuzzers'
.//run_fuzzer.py 000-init-db
Traceback (most recent call last):
  File ".//run_fuzzer.py", line 703, in <module>
    sys.exit(main(sys.argv))
  File ".//run_fuzzer.py", line 455, in main
    part = get_part_information(db_root, os.environ['XRAY_PART'])
  File "/tmpfs/src/github/symbiflow-prjxray-presubmit-db-zynq7/prjxray/util.py", line 43, in get_part_information
    assert os.path.isfile(filename), \
AssertionError: Mapping file /tmpfs/src/github/symbiflow-prjxray-presubmit-db-zynq7/database/zynq7/mapping/parts.yaml does not exists
Makefile:87: recipe for target '000-init-db/run.xc7z020clg484-1.ok' failed
make[1]: *** [000-init-db/run.xc7z020clg484-1.ok] Error 1
make[1]: Leaving directory '/tmpfs/src/github/symbiflow-prjxray-presubmit-db-zynq7/fuzzers'
Makefile:170: recipe for target 'db-zynq7' failed
make: *** [db-zynq7] Error 2


[ID: 1790215] Build finished after 135 secs, exit value: 1


Warning: Permanently added 'localhost' (ECDSA) to the list of known hosts.
[16:20:24] Collecting build artifacts from build VM
Build script failed with exit code: 1
```

litghost commented:
> Looks like this is failing due to the `Cleaning out current database`?
> 
> ```
> ========================================
> Cleaning out current database
> ----------------------------------------
> rm -f zynq7/*.csv zynq7/*.db zynq7/*.json zynq7/*.yaml zynq7/*.fasm
> rm -f zynq7/timings/*.sdf
> rm -rf zynq7/xc7*/
> ----------------------------------------
> 
> ========================================
> Running Database build
> ----------------------------------------
> make --dry-run
> source settings/zynq7.sh && make -C fuzzers
> make[1]: Entering directory '/tmpfs/src/github/symbiflow-prjxray-presubmit-db-zynq7/fuzzers'
> .//run_fuzzer.py 000-init-db
> Traceback (most recent call last):
>   File ".//run_fuzzer.py", line 703, in <module>
>     sys.exit(main(sys.argv))
>   File ".//run_fuzzer.py", line 455, in main
>     part = get_part_information(db_root, os.environ['XRAY_PART'])
>   File "/tmpfs/src/github/symbiflow-prjxray-presubmit-db-zynq7/prjxray/util.py", line 43, in get_part_information
>     assert os.path.isfile(filename), \
> AssertionError: Mapping file /tmpfs/src/github/symbiflow-prjxray-presubmit-db-zynq7/database/zynq7/mapping/parts.yaml does not exists
> Makefile:87: recipe for target '000-init-db/run.xc7z020clg484-1.ok' failed
> make[1]: *** [000-init-db/run.xc7z020clg484-1.ok] Error 1
> make[1]: Leaving directory '/tmpfs/src/github/symbiflow-prjxray-presubmit-db-zynq7/fuzzers'
> Makefile:170: recipe for target 'db-zynq7' failed
> make: *** [db-zynq7] Error 2
> 
> 
> [ID: 1790215] Build finished after 135 secs, exit value: 1
> 
> 
> Warning: Permanently added 'localhost' (ECDSA) to the list of known hosts.
> [16:20:24] Collecting build artifacts from build VM
> Build script failed with exit code: 1
> ```

@dnltz hasn't added the YAML files to this PR yet, which is why it failed.

dnltz commented:
@litghost @mithro - Sorry for the chaos! My understand was the database is cloned completely from prjxray-db. However, I added exceptions to the gitignore file for the yaml files and placed them in database/<family>/mapping and settings/<family>.
Since the files are in the project now, kokoro should run :)

litghost reviewed (approved):
LGTM, waiting to see CI behavior before merging.  Thanks!

litghost commented:
Initial CI run looks good, just need to waiting for it to finish to see if we have any suprise.
